File: backend/policy-service/src/main/java/com/insurance/policyservice/service/PolicyService.java
package com.insurance.policyservice.service;

import com.insurance.policyservice.client.CustomerClient;
import com.insurance.policyservice.dto.CustomerDto;
import com.insurance.policyservice.dto.PolicyWithCustomerDto;
import com.insurance.policyservice.model.Policy;
import com.insurance.policyservice.repository.PolicyRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class PolicyService {
    @Autowired
    private PolicyRepository policyRepository;

    @Autowired
    private CustomerClient customerClient;

    public List<Policy> getAllPolicies() {
        return policyRepository.findAll();
    }

    public Optional<Policy> getPolicyById(Long id) {
        return policyRepository.findById(id);
    }

    public Policy createPolicy(Policy policy) {
        return policyRepository.save(policy);
    }

    public Policy updatePolicy(Long id, Policy policyDetails) {
        Policy policy = policyRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Policy not found with id: " + id));
        
        policy.setType(policyDetails.getType());
        policy.setStartDate(policyDetails.getStartDate());
        policy.setEndDate(policyDetails.getEndDate());
        policy.setCoverageAmount(policyDetails.getCoverageAmount());
        
        return policyRepository.save(policy);
    }

    public void deletePolicy(Long id) {
        Policy policy = policyRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Policy not found with id: " + id));
        
        policyRepository.delete(policy);
    }

    public List<Policy> getPoliciesByCustomerId(Long customerId) {
        return policyRepository.findByCustomerId(customerId);
    }

    public PolicyWithCustomerDto getPolicyWithCustomer(Long id) {
        Policy policy = policyRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Policy not found with id: " + id));
        
        CustomerDto customer = customerClient.getCustomerById(policy.getCustomerId());
        
        return new PolicyWithCustomerDto(policy, customer);
    }

    public List<PolicyWithCustomerDto> getAllPoliciesWithCustomers() {
        List<Policy> policies = policyRepository.findAll();
        
        return policies.stream()
                .map(policy -> {
                    CustomerDto customer = customerClient.getCustomerById(policy.getCustomerId());
                    return new PolicyWithCustomerDto(policy, customer);
                })
                .collect(Collectors.toList());
    }
}

File: frontend/src/services/policyService.ts
import api from "./api"
import type { Policy, PolicyWithCustomer } from "../types/Policy"

const POLICIES_URL = "/policies"

const getAllPolicies = async (): Promise<Policy[]> => {
  const response = await api.get<Policy[]>(POLICIES_URL)
  return response.data
}

const getPolicyById = async (id: number): Promise<Policy> => {
  const response = await api.get<Policy>(`${POLICIES_URL}/${id}`)
  return response.data
}

const getPolicyWithCustomer = async (id: number): Promise<PolicyWithCustomer> => {
  const response = await api.get<PolicyWithCustomer>(`${POLICIES_URL}/with-customer/${id}`)
  return response.data
}

const getAllPoliciesWithCustomers = async (): Promise<PolicyWithCustomer[]> => {
  const response = await api.get<PolicyWithCustomer[]>(`${POLICIES_URL}/all-with-customers`)
  return response.data
}

const getPoliciesByCustomerId = async (customerId: number): Promise<Policy[]> => {
  const response = await api.get<Policy[]>(`${POLICIES_URL}/customer/${customerId}`)
  return response.data
}

const createPolicy = async (policy: Policy): Promise<Policy> => {
  const response = await api.post<Policy>(POLICIES_URL, policy)
  return response.data
}

const updatePolicy = async (id: number, policy: Policy): Promise<Policy> => {
  const response = await api.put<Policy>(`${POLICIES_URL}/${id}`, policy)
  return response.data
}

const deletePolicy = async (id: number): Promise<void> => {
  await api.delete(`${POLICIES_URL}/${id}`)
}

const policyService = {
  getAllPolicies,
  getPolicyById,
  getPolicyWithCustomer,
  getAllPoliciesWithCustomers,
  getPoliciesByCustomerId,
  createPolicy,
  updatePolicy,
  deletePolicy,
}

export default policyService

